clang/test/CodeGenSYCL/sycl-device-static-init.cpp

```diff
@@ -0,0 +1,43 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -disable-llvm-passes %s -emit-llvm -o -  | FileCheck %s
+// Test that static initializers do not force the emission of globals on sycl device
+
+// CHECK: %struct._ZTS16RegisterBaseInit.RegisterBaseInit = type { i8 }
+// CHECK-NOT: $_ZN8BaseInitI12TestBaseTypeE15s_regbase_ncsdmE = comdat any
+// CHECK: $_ZN8BaseInitI12TestBaseTypeE3varE = comdat any
+// CHECK: @_ZN8BaseInitI12TestBaseTypeE9s_regbaseE = {{.*}} global %struct._ZTS16RegisterBaseInit.RegisterBaseInit
+// CHECK-NOT: @_ZN8BaseInitI12TestBaseTypeE15s_regbase_ncsdmE = weak_odr addrspace(1) global %struct._ZTS16RegisterBaseInit.RegisterBaseInit zeroinitializer, comdat, align 1
+// CHECK: @_ZN8BaseInitI12TestBaseTypeE3varE = weak_odr addrspace(1) constant i32 9, comdat, align 4
+// CHECK: @llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 65535, void ()* @__cxx_global_var_init, i8* addrspacecast (i8 addrspace(1)* getelementptr inbounds (%struct._ZTS16RegisterBaseInit.RegisterBaseInit, %struct._ZTS16RegisterBaseInit.RegisterBaseInit addrspace(1)* @_ZN8BaseInitI12TestBaseTypeE9s_regbaseE, i32 0, i32 0) to i8*) }]
```

Comment:
Just one more question, why we still emit `llvm.global_ctors`, AFAIK this is array of constructors for globals and they are called when module is loaded. https://llvm.org/docs/LangRef.html#id1395
In SYCL, global variables which can be used in device code must be const/constexpr and zero-initialized/const-initialized. I'm not sure that we should emit any constructors for global variables at all.

Reply:
It is being used for initializing the const static data member of the class.  It seems to be needed.   A version without this broke the fpga pipes in testing a couple days back.